### PR TITLE
[DependencyInjection] Fixed resolving of service configurators containing Definition objects

### PR DESCRIPTION
**Pull request by webmozart**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14834 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -981,9 +981,7 @@ public function createService(Definition $definition, $id, $tryProxy = true)
         }
 
         if ($callable = $definition->getConfigurator()) {
-            if (is_array($callable)) {
-                $callable[0] = $callable[0] instanceof Reference ? $this->get((string) $callable[0]) : $parameterBag->resolveValue($callable[0]);
-            }
+            $callable = $this->resolveServices($parameterBag->resolveValue($callable));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: this now allows expressions too, which looks wrong to me

Comment (webmozart) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: why not be consistent with the rest?

Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Expressions are not supported in configurators in other places. They are only supported in arguments. So you don't introduce consistency but inconsistency (and using an expression for the service instance of the callable does not make sense either btw)

Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: I agree with @stof.

Comment (webmozart) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Alright. What solution do you propose then? Duplicate `resolveServices()` and remove expression support from the duplicate? Any naming suggestions?

Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: well, we don't need to support the case of arrays either for configurators first element. So this is just a matter of adding the support of Definition object in the current code (which already supports references)

Comment (webmozart) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Yes, but that means we still have duplicated and potentially buggy code. If somebody adds functionality to resolveServices(), they need to think about changing this code too (like happened before). IMO this should be fixed.

**Comment by webmozart** [comment]
I updated the PR to match @stof's comments. I don't like this solution however, because the duplicated code in `createService()` and `resolvedServices()` bears potential for new bugs.

**Comment by fabpot** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @webmozart.

**Pull request merged**